rocAL/source/readers/file_source_reader.cpp

```diff
@@ -41,10 +42,32 @@ FileSourceReader::FileSourceReader() {
 }
 
 unsigned FileSourceReader::count_items() {
-    if (_loop)
-        return _file_names.size();
-
-    int ret = ((int)_file_names.size() - _read_counter);
+    int ret = 0;
+    if (_shard_size == -1) {                                     // When shard_size is set to -1, The shard_size variable is not used
+        if (_loop) return largest_shard_size_without_padding();  // Return the size of the largest shard amongst all the shard's size
+        int size = std::max(largest_shard_size_without_padding(), _batch_count);
+        ret = (size - _read_counter);
+        // Formula used to calculate - [_last_batch_padded_size = _batch_count - (_shard_size % _batch_count) ]
+        // Since "size" doesnt involve padding - we add the count of padded samples to the number of remaining elements
+        // which equals to the shard size with padding
+        if (_last_batch_info.first == RocalBatchPolicy::PARTIAL || _last_batch_info.first == RocalBatchPolicy::FILL) {
+            ret += _last_batch_padded_size;
+        } else if (_last_batch_info.first == RocalBatchPolicy::DROP &&
+                   _last_batch_info.second == true) {  // When pad_last_batch_repeated is False - Enough
+                                                       // number of samples would not be present in the last batch - hence
+                                                       // dropped by condition handled in the loader
+            ret -= _batch_count;
+        }
+    } else if (_shard_size > 0) {
+        auto largest_shard_size_with_padding =
+            _shard_size + (_batch_count - (_shard_size % _batch_count));  // The shard size used here is padded
+        if (_loop)
+            return largest_shard_size_with_padding;
+        int size = std::max(largest_shard_size_with_padding, _batch_count);
+        ret = (size - _read_counter);
+        if (_last_batch_info.first == RocalBatchPolicy::DROP)  // The shard size is padded at the beginning of the condition, hence dropping the last batch
+            ret -= _batch_count;
```

Comment:
the dropping has to happen at the caller. So returning -= _batch_size is not recommended

Reply:
Since we are using the padded shard_size at the beginning, for dropping, we will have to subtract with the shard_size to maintain the batch count.